kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -271,48 +272,271 @@ void KvbAppFilter::readBlockRange(BlockId block_id_start,
+      for (uint64_t i = eg_state->private_offset; i < std::min(eg_state->private_offset + kBatchSize, private_end + 1);
+           ++i) {
+        uint64_t global_eg_id =
+            getLatestFromTagTableInStorage(client_id + "#" + concordUtils::toBigEndianStringBuffer(i));
+        private_event_group_ids.emplace_back(global_eg_id);
+      }
+    }
+    // populate event_group_id_batch with only public event group ids iff one of the following occurs -\
+    // 1. No private event groups exist in storage \
+    // 2. The first global event group id in private_event_group_ids is greater than the last global event group id public_event_group_ids\
+    // Vice versa is true for populate event_group_id_batch with only private event group ids \
+    // Note: each private_event_group_ids and public_event_group_ids hold a sorted list of global event group ids at all times
+    if (!public_event_group_ids.empty() &&
+        (private_event_group_ids.empty() || (public_event_group_ids.back() <= private_event_group_ids.front()))) {
+      for (auto public_event_group_id : public_event_group_ids) {
+        eg_state->event_group_id_batch.emplace_back(public_event_group_id);
+      }
+      ConcordAssert(eg_state->event_group_id_batch.size() <= kBatchSize);
+      // increment the offset
+      eg_state->public_offset += eg_state->event_group_id_batch.size();
+      LOG_DEBUG(logger_,
+                "Updated public_offset: " << eg_state->public_offset
+                                          << " public_event_group_ids size: " << public_event_group_ids.size());
+    } else if (!private_event_group_ids.empty() &&
+               (public_event_group_ids.empty() || (private_event_group_ids.back() <= public_event_group_ids.front()))) {
+      for (auto private_event_group_id : private_event_group_ids) {
+        eg_state->event_group_id_batch.emplace_back(private_event_group_id);
+      }
+      ConcordAssert(eg_state->event_group_id_batch.size() <= kBatchSize);
+      // increment the offset
+      eg_state->private_offset += eg_state->event_group_id_batch.size();
+      LOG_DEBUG(logger_,
+                "Updated private_offset: " << eg_state->private_offset
+                                           << " private_event_group_ids size: " << private_event_group_ids.size());
+    } else if (!public_event_group_ids.empty() && !private_event_group_ids.empty()) {
+      LOG_DEBUG(logger_, "Both public and private event groups found in the batch");
+      std::merge(public_event_group_ids.begin(),
+                 public_event_group_ids.end(),
+                 private_event_group_ids.begin(),
+                 private_event_group_ids.end(),
+                 std::back_inserter(eg_state->event_group_id_batch));
+      // We cannot return the full merge
+      // E.g.: a=[1,2,3,4,5] b=[6,7,8,9]
+      if (eg_state->event_group_id_batch.size() > kBatchSize) {
+        eg_state->event_group_id_batch.erase(eg_state->event_group_id_batch.begin() + kBatchSize,
```

Comment:
Curious, why using `resize` is better?

Reply:
erase(const_iterator __first, const_iterator __last)
      {
        const auto __beg = begin();
        const auto __cbeg = cbegin();
        return _M_erase(__beg + (__first - __cbeg), __beg + (__last - __cbeg));
      }


        void
      resize(size_type __new_size)
      {
        if (__new_size > size())
          _M_default_append(__new_size - size());
        else if (__new_size < size())
          _M_erase_at_end(this->_M_impl._M_start + __new_size);
      }

So we see that resize() [in one of the implementation, i am not saying its in general] is using the fact that it has to delete from the end.
So it can call _M_erase_at_end() which is internally just some index switching. 
But in case of erase() it doesn't know which part of vector is getting erased. So it cannot do the optimisation, instead it calls _M_erase() which might also have to copy the items from the second index and place it in the first index.

So its preferable to call resize, since we can tell STL that this is something that need to be chopped off from end and STL can optimize if possible. But again in some STL implementation both erase and resize can be same.
Its the profiler that can decide in terms of performance. But in general its a good practice since by saying x.resize we are telling the intent that remove items from end and not from any other range within the vector.